### ISSUE TITLE
[1.2.0] Unable to get document FILENAME.HTML:

### ISSUE THREAD

**Issue by beaudesigns** [Polymer/polymer-build#194]
### Description
My build script simply attempts to build a single-file version of Polymer 2.0. There are no custom files imported.

After upgrading to 1.2.0, my Gulp build script fails to build my Polymer bundle. The message is
> Unable to get document polymer-bundle.html:

1.1.0 works as I would expect.

I tracked it down to `deps-index.js` inside of the `buildDepsIndex()` function. I realize this is a dependency of this project, so if I need to move this report, just let me know.

### Versions & Environment
- polymer-build: v1.2.0
- node: v7.9.0
- Operating System: OSX 10.12.4

#### Steps to Reproduce
My `polymer.json` file:
```
{
  "entrypoint": "index.html",
  "fragments": [
    "polymer-bundle.html"
  ],
  "sources": [],
  "extraDependencies": [],
  "builds": [
    {
      "name": "bundled",
      "bundle": true,
      "js": {
        "minify": true
      },
      "css": {
        "minify": true
      },
      "html": {
        "minify": true
      }
    }
  ],
  "lint": {
    "rules": [
      "polymer-2-hybrid"
    ]
  }
}
```

My `polymer-bundle.html` file:
```
<!-- polymer library -->
<link rel="import" href="bower_components/polymer/polymer.html">

<!-- extra polymer behaviors -->
<link rel="import" href="bower_components/polymer/lib/mixins/gesture-event-listeners.html">
```

#### Expected Results
A bundled version of Polymer and its imported files.

#### Actual Results
An error is thrown and my build fails.

**Comment by usergenic** [comment]
@beaudesigns is the `polymer-bundle.html` fragment present in the gulp stream that you're piping into Bundler?

Also, can you post your gulpfile or point to a repo or gist?

**Comment by beaudesigns** [comment]
Sure, the relavant portions of my gulp file are:

```
const mergeStream = require('merge-stream');
const PolymerProject = require('polymer-build').PolymerProject;
const project = new PolymerProject(require('./polymer.json'));

gulp.task('controls:polymer', false, function () {
      return mergeStream(project.sources(), project.dependencies())
          .pipe(project.bundler())
          .pipe(gulp.dest('build/bundled'));
    });
```

**Comment by mraerino** [comment]
Happening for me as well. Except the build doesn't fail, but outputs non-bundled elements.

**Comment by krumware** [comment]
Hi folks, had the same issue here. unbundled version was being output fine, but not the bundled version. Our build worked great with polymer-build 1.1.0, but not 1.2.0+

After toying with the packages, I modified node_modules/polymer-build/analyzer.js by adding a console log to onSourcesStreamComplete
```
onSourcesStreamComplete() {
        // Emit an error if there are missing source files still deferred. Otherwise
        // this would cause the analyzer to hang.
        for (const filePath of this.loader.deferredFiles.keys()) {
            if (this.config.isSource(filePath)) {
                const err = new Error(`Not found: ${filePath}`);
                // ADDED THIS LOG
               console.log(err);
                this.loader.rejectDeferredFile(filePath, err);
                return;
            }
        }
        // Set sourceFilesLoaded so that future files aren't accidentally deferred
        this.sourceFilesLoaded = true;
    }
```
This helped us find some misnamed paths, and a missing file or two or three. It seems the errors are not bubbling up to gulp.  You can also comment out the `return;` if you would like it to return multiple errors instead of quitting immediately on the first error.

Seems the updated polymer-build is less forgiving about missing file references in those link imports.

**Comment by usergenic** [comment]
@beaudesigns you may be able to identify your issues in gulp flow this way, since pipe will not propagate errors:

```js
gulp.task('controls:polymer', false, function () {
      return mergeStream(project.sources(), project.dependencies()).on('error', (e) => console.log(e))
          .pipe(project.bundler()).on('error', (e) => console.log(e))
          .pipe(gulp.dest('build/bundled'));
    });
```

**Comment by beaudesigns** [comment]
@usergenic I tried that, and it actually doesn't ever hit the error handlers. It just logs what I got before
> Unable to get document polymer-bundle.html:

I verified that that doesn't come from the error handlers. No `error` event.

**Comment by usergenic** [comment]
So I finally encountered this today and the issue was that the `Document` object returned from the analysis being used in the deps-index.js point does not have the same constructor as is expected, which happens when multiple versions of the `polymer-analyzer` are present in the node_modules tree.

The code should probably be updated to not use `instanceof` to type-check the result here, but in the short term, you should probably just verify that only one version of analyzer is in your tree:
```
$ find . | grep analyzer\$
node_modules/polymer-analyzer
```

**Comment by usergenic** [comment]
@beaudesigns let me know if this helps you.

**Comment by beaudesigns** [comment]
@usergenic I have three instances of `polymer-analyzer`.

```
./node_modules/polymer-analyzer
./node_modules/polymer-build/node_modules/polymer-analyzer
./node_modules/polymer-bundler/node_modules/polymer-analyzer
```

I include `polymer-analyzer` in my `package.json` and use it in my gulpfile to extract some build information from components.

Both `polymer-analyzer` and `polymer-build` are installed as "latest" in my `package.json` file when it breaks. To get things working as expected, I'm installing `polymer-build#1.1.0`.

Thanks for your info!

**Comment by web-padawan** [comment]
Also encountered this issue, reproduced in my [sandbox](https://github.com/web-padawan/polymer-sandbox/tree/build-warnings), `build-warnings` branch.

**Comment by usergenic** [comment]
The problem is that if there are multiple versions of polymer-analyzer, code that checks for documents to be "instanceof Document" fail and the symptom is that the document was not returned from the analysis.

The most recent versions of polymer build and polymer bundler use ^2.0.0 of polymer analyzer instead of pegging it to a specific prerelease version.  I would suggest upgrading to latest build and bundler, if you specifically include it in your package.json and blow out your node_modules folder with a fresh install to insure polymer-analyzer only appears once in the tree.

You want it to look like this:
```
$ find node_modules | grep polymer-analyzer/package.json | xargs grep version
  "version": "2.0.1"
```

**Comment by usergenic** [comment]
web-component-tester/polyserve will also bring in the dependency.

We're looking at updating the code to use something less version-finicky than "instanceof" but upgrading to latest version of the tools should enable them to all select the latest versions and eliminate duplicates in the node_modules tree.

**Comment by beaudesigns** [comment]
I got mine working, I'm using `yarn`, so I had to:

- Set my `package.json` versions to the correct version. In my case `latest`.
- Run `yarn` to update my dependencies.
- Run `yarn upgrade` to update any sub dependencies.

After that it's working great. Thanks for all your help and patience @usergenic.

**Comment by usergenic** [comment]
I'm going to close this issue.  If anyone else is still having it, please re-open (@web-padowan)

**Comment by usergenic** [comment]
@web-padawan  (i always typo that sorry) see above.

**Comment by gdevacc12** [comment]
With polymer-cli v1.5.7 I had the same problem.

> Bundler was given a different version of polymer-analyzer than expected.  Please ensure only one version of polymer-analyzer present in node_modules folder:
> $ npm ls polymer- analyzer

I'm not sure if this was a problem: with my specific config (some legacy package), the resulting from an vanilla install of polymer-cli with yarn (indicating some general package config issue), or if this is how the package dependencies are supposed to be, BUT in my case there were multiple versions of polymer-analyzer installed.

I found the multiple installs using:  find ~ | grep analyzer\$

From ~/.config/yarn/global/node_modules (Ubuntu 16.04), polymer-analyzer:
2.4.1 was installed in ./polymer-cli/node_modules/polymer-analyzer
2.4.1 was installed in ./polymer-bundler/node_modules/polymer-analyzer
2.2.1 was installed in ./polymer-analyzer
Workaround actions:
moved 2.1.1 (~/.config/yarn/global/node_modules/polymer-analyzer) to ~/polymer-analyzer-old (just in case I broke something)
moved ./polymer-cli/node_modules/polymer-analyzer to ./polymer-analyzer
removed ./polymer-bundler/node_modules/polymer-analyzer (as it was also 2.4.1)

polymer build appears to run OK and there were no warning messages.
Warning: I'm not sure of the integrity of the package dependencies, particularly if there's an update to polymer-cli but at least this appears to be at least workaround.
